### ISSUE TITLE
Introduce OutputType

### ISSUE THREAD

**Issue by lipkau** [PSJira/PSJira#96]
The declaration of `[OutputType()]` helps the intellisense of powershell (eg: in _vs code_)

So I would like for every public function of the module to have that attribute.

**Comment by lipkau** [comment]
@replicaJunction do you have an opinion on this?

**Comment by replicaJunction** [comment]
I agree this is a good practice. The biggest hurdle with this, and the reason I didn't already implement it, is that objects in the PSJira namespace aren't real .NET classes - they're just PSCustomObjects with a custom typename added. There's no class definition for things like IntelliSense to really be useful.

I toyed with the idea of creating classes for them via C# (either with a compiled library or just using Add-Type) but this breaks some of the functionality in the ConvertTo-Jira* functions where it adds additional custom properties to the object. We could do something like create a base class, then use Add-Member to add custom items to it.

I haven't tested this extensively, so if OutputType is still useful even without real classes, I don't see any reason not to add it.

**Comment by lipkau** [comment]
That is a good point.

if custom type can be used with customfields being added with `Add-Type`, would that be a direction you were willing to go?

benefits could be:
* the custom objectt have methods to add custom fields (as they will not be available when "newing up" an object
```posh
$Issue = New-Object -Type PSJira.Issue
$Issue.Summary = "Lorem Ipsum"
$Issue.SetField("Epic Link", "FOO-1000")
```
* custom object could have a method that saves us the hassle of setting up the json structure (like `.toString()`)
* we could make a function that would validate a custom object against `Get-JiraIssueCreateMetadata` and `Get-JiraIssueEditMetadata`
```posh
New-Object -Type PSJira.Issue | Validate-JiraIssue -Action "create" | New-JiraIssue
```

**Comment by replicaJunction** [comment]
Something like this?

```csharp
namespace PSJira {
    public class Issue {
        public string Key { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        # ...
    }
}
```

Then, in ConvertTo-JiraIssue....
```powershell
# Create issue object...assume $props contains all standard Jira fields that would be in the class def
$obj = New-Object -TypeName PSJira.Issue -Properties $props

# Process extra properties not in the class definition
# $i here is coming from $InputObject
$extraFields = $i.fields.PSObject.Properties | Where-Object -FilterScript { $_.Name -notin $props.Keys }
foreach ($f in $extraFields)
{
    $obj | Add-Member -MemberType NoteProperty -Name $f.name -Value $f.Value
}
```

When I messed with this stuff before, I didn't really understand how powerful Add-Member can be. I think that would be a great compromise - that way, we get some basic code completion (even if it doesn't include custom fields and such), and there's no real impact to the user's experience. We _might_ even be able to have users create .type.ps1xml files if they want to extend PowerShell's definition of that object to include their custom properties.

I'm also really digging the idea of a Validate-JiraIssue function. That could get rid of some of all the API calls New-JiraIssue uses to make sure the object is valid before it actually creates the issue.

I'm having trouble understanding how the .SetField() method you referenced above would work, unless there was just a dictionary / hashtable of fields in the class definition. In that case, I'd consider that a breaking change - currently you can reference ```$issue.customfield_10010```, but in that scenario, you'd need to reference ```$issue.fields.customfield_10010```. Not a huge inconvenience, but it could break production scripts.
